### ISSUE TITLE
Uri to specific line doesn't work

### ISSUE THREAD

**Issue by tomi** [Microsoft/vscode#12562]
- VSCode Version: 1.5.1
- OS Version: 10.11.6 (15G1004)

Steps to Reproduce:

1. Create an extension that uses `vscode.previewHtml` command to open an URI that is registered to `TextDocumentContentProvider`
2. Generate a document that contains a link to a file URI to a specific line. I've noticed the file needs to be in the current project that is open. If it's not, the bug doesn't appear.
3. The link doesn't work when it has also a line reference, without it it works.

For example I was able to reproduce the issue with the extension below. Execute the command when there's a file open in the project. The URI to the file to a specific line doesn't work, but the URI to the same file without line specified it works.

```javascript
"use strict";

const vscode = require("vscode");

const scheme = "test-scheme";
const eventEmitter = new vscode.EventEmitter();

const provider = {
    scheme,

    provideTextDocumentContent: function(uri) {
        const file = vscode.window.activeTextEditor.document.uri.toString();
        const urlWithLine = `${ file }#L10`;
        const urlWoLine = `${ file }`;
        return `
<a href="${ urlWithLine }">Link to file with line, doesn't work</a><br/><br/>
<a href="${ urlWoLine }">Link to same file without line, works</a>
`;
    },

    onDidChange: function() {
        return eventEmitter.event;
    },

    update: function(uri) {
        eventEmitter.fire(uri);
    }
};

function exec() {
    const uri = vscode.Uri.parse(`${ scheme }://test/test`);
    vscode.commands.executeCommand("vscode.previewHtml", uri);
}

function activate(context) {
    context.subscriptions.push(vscode.commands.registerCommand("testExt.exec", exec));

    context.subscriptions.push(vscode.workspace.registerTextDocumentContentProvider(scheme, provider));
}

exports.activate = activate;
```

![vscode-bug](https://cloud.githubusercontent.com/assets/10324676/18810970/879d1388-82ac-11e6-9e65-a252cfdeabb2.gif)

**Comment by jrieken** [comment]
@kieferrm This is a regression from https://github.com/Microsoft/vscode/commit/f165bc60395537ed9bd9a56880feb81064329e92 and I think I finally need to understand you guys did there ;-)

**Comment by kieferrm** [comment]
The code tries to determine if the anchor target is within the displayed document and if so to scroll to it. For that it uses the webView's baseUrl. The baseUrl, however,  is set to the workspace root not the url that is being displayed in the html preview. As this is the only use/purpose of the baseUrl we should change the baseUrl to be the uri of the resource being displayed in the html preview. This could be achieved by changing `HtmlPreviewPart:setInput` along the lines of 
```
	public setInput(input: EditorInput, options: EditorOptions): TPromise<void> {

		if (this.input === input && this._hasValidModel()) {
			return TPromise.as(undefined);
		}

		this._model = undefined;
		this._modelChangeSubscription.dispose();

		if (!(input instanceof HtmlInput)) {
			return TPromise.wrapError<void>('Invalid input');
		}

		return super.setInput(input, options).then(() => {
			let resourceUri = (<HtmlInput>input).getResource();
			return this._editorService.resolveEditorModel({ resource: resourceUri }).then(model => {
				if (model instanceof BaseTextEditorModel) {
					this._model = model.textEditorModel;
				}
				if (!this._model) {
					return TPromise.wrapError<void>(localize('html.voidInput', "Invalid editor input."));
				}
				this._modelChangeSubscription = this._model.onDidChangeContent(() => this.webview.contents = this._model.getLinesContent());
				this.webview.baseUrl = resourceUri.toString(true);
				this.webview.contents = this._model.getLinesContent();
			});
		});
	}
```

and by making sure the web view is creating or updating the base element accordingly.

**Comment by jrieken** [comment]
Unsure if that breaks cases in which folks have server relative paths (we have seen things like that with the extensions preview). We should consider using `window.location` for the in-page navigation.

**Comment by jrieken** [comment]
fixed with #12966
